Source/DafnyTestGeneration/README.md

```diff
@@ -0,0 +1,149 @@
+- If you wish to test a particular method rather than all the methods in a file, you can specify such a method with the `/testTargetMethod` command line argument and providing the fully qualified method name.
+- If you are using `/testTargetMethod` and would like to inline methods that are called from the method of interest, you can do so by setting `/testInlineDepth` to something larger than zero (zero is the default). The `/verifyAllModules` argument might also be relevant if the methods to be inlined are defined in included files.
+- To deal with loops, you should use `/loopUnroll` and also `/testSeqLengthLimit`. The latter argument adds an axiom that limits the length of any sequence to be no greater than some specified value. This restriction can be used to ensure that the number of loop unrolls is sufficient with respect to the length of any input sequence but it can also cause the program to miss certain corner cases.
+- The`/testMode:DeadCode` argument will make Dafny identify potential dead code in the specified file. Note that false negatives are possible if `/loopUnroll` is not used. False positives are also possible for a variety of reasons, such as `/loopUnroll` being assigned not high enough value.
+
+## How to Run Tests?
+
+To run the tests, you would first need to compile the code to some other language. Once this is done, the main difficulty is to mock the arguments that the test methods take. The following code (to be inserted into the class with the tests) should do the job in Java:
+
+```java
+public static void main(String[] args) throws Exception {
+  for (Method method: CLASS_WITH_TESTS.class.getDeclaredMethods()) {
+    if (!method.getName().startsWith("test"))
+      continue;
+    Object result = method.invoke(null, mockParameters(method));
+    String resultString = result == null ? "null" : result.toString();
+    System.out.println(method.getName() + " " + resultString);
+  }
+}
+
+
+public static Object[] mockParameters(Method method) {
+    Object[] parameters = null;;
+    try {
+        Class<?>[] types = method.getParameterTypes();
+        parameters = new Object[types.length];
+        for (int i = 0; i < types.length; i++)
+            parameters[i] = types[i].getConstructor().newInstance();
+    } catch (Exception e) {
+        e.printStackTrace();
+    }
+    return parameters;
+}
+```
+
+## Example
+
+Suppose you have a file called `Char.dfy` with the following code:
+```
+module Char {
+  method compareToB(c: char) returns (i: int) {
+    if (c == 'B') {
+        return 0;
+    } else if (c > 'B') {
+        return 1;
+    } else {
+        return -1;
+    }
+  }
+}
+```
+Running test generation like this:
+
+```dafny /definiteAssignment:3 /testMode:Block Char.dfy ```
+
+Will give the following list of tests (tabulation added manually):
+```
+include "Char.dfy"
+module CharUnitTests {
+  import Char
+  method test0() returns (r0:int)  {
```

Comment:
Would `{:test}` compile successfully if used on methods that take parameters? (this is necessary to mock objects as outlined in "How to run tests" section in the [README](https://github.com/Dargones/dafny/tree/PatchesToCounterExamples/Source/DafnyTestGeneration#readme)). I agree that it would be great to use the annotation but I am wondering if it would be better to create a separate PR about extending the way the compilers treat `{:test}` before using it in this way.

Reply:
No, it wouldn't work if the methods take parameters without more work, even in the C# case. `[Fact]` only works if there are no parameters. `[Theory]` can be used to support parameterized tests, but you'd have to supply a separate `DataAttribute` of some kind to provide the mocked object arguments.

No worries if you want to ignore this suggestion for now - it may be better to follow up on it once Dafny's support for tests is fleshed out more.